components/raftstore/src/store/peer_storage.rs

```diff
@@ -436,16 +444,41 @@ where
     /// Gets a snapshot. Returns `SnapshotTemporarilyUnavailable` if there is no
     /// unavailable snapshot.
     pub fn snapshot(&self, request_index: u64, to: u64) -> raft::Result<Snapshot> {
+        if self.peer.as_ref().unwrap().is_witness {
+            // witness could be the leader for a while, do not generate snapshot now
+            return Err(raft::Error::Store(
+                raft::StorageError::SnapshotTemporarilyUnavailable,
+            ));
+        }
+
+        let for_witness = find_peer_by_id(&self.region, to).unwrap().is_witness;
```

Comment:
I think may need to add a judgment if the peer is in `wait_data_peers` See: https://github.com/tikv/tikv/pull/12972/files#r1000411516

Reply:
We can validate the snapshot on receiving and reject it to resend a new snapshot.